models.py

```diff
@@ -626,28 +793,68 @@ def design_with_fullmpnn(batched_coords, seq_mask):
                 xt.requires_grad = True
 
             # Run denoising network
-            run_mpnn = not sidechain_mode or i > begin_mpnn_step
-            x0, s_logprobs, x_self_cond, s_self_cond = self.forward(
-                noisy_coords=xt,
-                noise_level=sigma,
-                seq_mask=seq_mask,
-                residue_index=residue_index,
-                struct_self_cond=x_self_cond,
-                struct_crop_cond=crop_cond_coords,
-                seq_self_cond=s_self_cond,
-                run_mpnn_model=run_mpnn,
-            )
+            run_mpnn = sidechain_mode and i > begin_mpnn_step
+            use_mpnn = sidechain_mode and i > begin_mpnn_step
+
+            ### Settings for deterministic & stochastic samplers
+            if not sidechain_mode:
+                if gamma > 0:  # backbone stochastic sampler, karras scheme
+                    sigma_hat = sigma + gamma * sigma
+                    sigma_delta = torch.sqrt(sigma_hat**2 - sigma**2)
+                    noiser_xt = xt + utils.expand(
+                        sigma_delta, xt
+                    ) * noise_scale * torch.randn_like(xt).to(xt)
+                    xt_hat = noiser_xt * utils.expand(seq_mask, noiser_xt)
+
+                    xt_hat *= bb_atom_mask[..., None]
+                    x0, s_logprobs, x_self_cond, s_self_cond = d_theta(
+                        xt_hat,
+                        sigma_hat,
+                        x_self_cond,
+                        s_self_cond,
+                        crop_cond_coords,
+                        run_mpnn,
+                        use_mpnn,
+                    )
+
+                else:  # backbone deterministic sampler, karras scheme
+                    x0, s_logprobs, x_self_cond, s_self_cond = d_theta(
+                        xt,
+                        sigma,
+                        x_self_cond,
+                        s_self_cond,
+                        crop_cond_coords,
+                        run_mpnn,
+                        use_mpnn,
+                    )
+
+            else:  # all-atom stochastic & deterministic samplers, Euler-Maruyama scheme
+                x0, s_logprobs, x_self_cond, s_self_cond = d_theta(
```

Comment:
do we also need to add the same fix (predict from xt_hat instead of xt when gamma > 0) as for backbone-only?

Reply:
I tried it before, but it didn't work. Maybe it's related to superposition scheme, so will try it later again.